include/kernel.h

```diff
@@ -4540,7 +4540,23 @@ extern void z_mem_pool_free_id(struct k_mem_block_id *id);
  */
 
 /**
- * @brief Allocate memory from heap.
+ * @brief Allocate memory from the heap with a specified alignment.
+ *
+ * This routine provides semantics similar to aligned_alloc(3); memory is
+ * allocated from the heap with a specified alignment. However, one minor
+ * difference is that k_aligned_alloc() accepts any non-zero @p size,
+ * wherase aligned_alloc(3) only accepts a @p size that is an integral
+ * multiple of @p align.
+ *
+ * @param align Alignment of memory requested (in bytes).
+ * @param size Amount of memory requested (in bytes).
+ *
+ * @return Address of the allocated memory if successful; otherwise NULL.
+ */
+extern void *k_aligned_alloc(size_t align, size_t size);
+
+/**
+ * @brief Allocate memory from the heap.
  *
  * This routine provides traditional malloc() semantics. Memory is
```

Comment:
Traditional malloc() semantics are _"The pointer returned if the allocation succeeds shall be suitably aligned so that it may be assigned to a pointer to any type of object and then used to access such an object in the space allocated [...]."_

It is not fully clear how this is guaranteed by the proposed of `k_malloc()` implementation. I would recommend either updating the description of `k_malloc()` here to state what the guaranteed alignment is, or add a comment in the implementation describing how it is achieved.

Reply:
> It is not fully clear how this is guaranteed by the proposed of `k_malloc()` implementation. I would recommend either 

Sure, let me clarify. It probably helps to reduce a call to `k_malloc(size)` before and after my change.

before ([ref](https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/mempool.c#L49)):
`k_malloc(size)`
-> `z_heap_alloc(heap, size)`
-> `k_heap_alloc(heap, size)`
-> `k_heap_aligned_alloc(heap, sizeof(void *), size, K_NO_WAIT)`

after:
`k_malloc(size)`
-> `k_aligned_alloc(sizeof(void *), size)`
-> `z_heap_aligned_alloc(heap, sizeof(void *), size)`
-> `k_heap_aligned_alloc(heap, sizeof(void *), size, K_NO_WAIT)`

As you can see, I have not made any significant modifications, nor do I propose to make any significant modifications, to the (underlying) implementation of `k_malloc()`.

As a generalization though, to satisfy the requirements for malloc, an implementation needs to return a pointer to a memory location that is aligned to the word size, which usually corresponds to the size of a `void *` on most architectures these days.